Add jni interface to use a binary hnsw index with faiss

heemin32 opened the pull request:
### Description
Add jni interface to use a binary hnsw index with faiss

SearchParameter is not supported with binary hnsw index in faiss library yet. Therefore, multi vector, pre filtering, and ef-search parameter passing won't work with binary index. Support of SearchParameter in binary index with faiss will be handled either by a PR in faiss repo or faiss custom patch in a separate PR.
 
### Issues Resolved
N/A
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

navneet1v commented on CHANGELOG.md:
```diff
@@ -27,3 +27,5 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Documentation
 ### Maintenance
 ### Refactoring
+* Add jni interface to use a binary hnsw index with faiss [#1747](https://github.com/opensearch-project/k-NN/pull/1747)
```
Should this be added in Enhancements or Features?

heemin32 replied:
Hmm. There is no enhancement or feature where end user can benefit from this PR as of now. Therefore I put it under refactoring.

navneet1v commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -81,6 +85,85 @@ bool isIndexIVFPQL2(faiss::Index * index);
 
+void knn_jni::faiss_wrapper::CreateBinaryIndex(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jintArray idsJ, jlong vectorsAddressJ, jint dimJ,
+                        jstring indexPathJ, jobject parametersJ) {
+    if (idsJ == nullptr) {
+        throw std::runtime_error("IDs cannot be null");
+    }
+
+    if (vectorsAddressJ <= 0) {
+        throw std::runtime_error("VectorsAddress cannot be less than 0");
+    }
+
+    if(dimJ <= 0) {
+        throw std::runtime_error("Vectors dimensions cannot be less than or equal to 0");
+    }
+
+    if (indexPathJ == nullptr) {
+        throw std::runtime_error("Index path cannot be null");
+    }
+
+    if (parametersJ == nullptr) {
+        throw std::runtime_error("Parameters cannot be null");
+    }
```
We should abstract the common code in separate functions. There is a lot of copy paste. Lets abstract things out.

heemin32 replied:
@navneet1v Will reduce duplicated logic.
@shatejas As I mentioned in other comment, If there should be a class for index creation related logic, it should be in a java side. Adding more logic in JNI side will make it harder to maintain.

heemin32 replied:
@shatejas On a second thought, maybe having a factory class might be good ad you suggested...

navneet1v commented on src/main/java/org/opensearch/knn/jni/FaissService.java:
```diff
@@ -49,6 +49,20 @@ class FaissService {
+     * @param indexPath path to save index file to
+     * @param parameters parameters to build index
+     */
+    public static native void createBinaryIndex(int[] ids, long vectorsAddress, int dim, String indexPath, Map<String, Object> parameters);
```
I don't think so we need this. Can we not use the parameters object to know if its a binary index or normal index.

heemin32 replied:
I believe such logic should be in Java side to make life easier. Also, the difference between binary index and non binary index is more than what it looks like at first glance. If we merge the two method, there will be lots of if/else and lots of different type of variables which will be used either by binary index or non binary index only.

shatejas replied:
> If we merge the two method, there will be lots of if/else and lots of different type of variables

I would want to understand this further. Shouldn't we be able to avoid/minimize if-else by introducing an interface? 

```
- Convert everything to cpp type upfront
- Call the factory which can decide on the type of index and execute the operations required
- delete references and return results
```

heemin32 replied:
The parameter seems same but actually it is not because what we have in vectorsAddress is different between binary index and non binary index.

Also, the factory might end up having two different method(like what we have now) to create index at the end anyway.

navneet1v commented on src/main/java/org/opensearch/knn/jni/FaissService.java:
```diff
@@ -90,6 +104,14 @@ public static native void createIndexFromTemplate(
+     * @param indexPath path to index file
+     * @return pointer to location in memory the index resides in
+     */
+    public static native long loadBinaryIndex(String indexPath);
```
same as above.

heemin32 replied:
same as above as well :)

navneet1v commented:
@heemin32 lets fix the GH actions too.

naveentatikonda commented on src/testFixtures/java/org/opensearch/knn/TestUtils.java:
```diff
@@ -327,10 +332,46 @@ private float[][] readQueries(String path) throws IOException {
+            Float median = flattenedVectors.get(flattenedVectors.size() / 2);
+
+            // Quantize + Packing for index data
+            indexBinaryData = new byte[indexData.vectors.length][(indexData.vectors[0].length + 7) / 8];
```
Can you add a comment explaining how you are quantizing and generating the data

heemin32 replied:
Quantization is obvious from the code I guess? `indexData.vectors[i][j] >= median ? 1 : 0`
Packing is also nothing special. Just packing 8 binary values into a single byte value.

heemin32 replied:
Let me add a comment.

shatejas reviewed (commented):
Reviewed this partly

heemin32 commented:
> @heemin32 lets fix the GH actions too.

Will do. So far, the error is coming from https://github.com/opensearch-project/k-NN/pull/1747/files#diff-892c5ae9158d35b9f5fa9b67d53d7373ca6d2bf04d1052add99c329d03645409R110 as index description is not passed to the method.

heemin32 commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -81,6 +85,85 @@ bool isIndexIVFPQL2(faiss::Index * index);
 // IndexIDMap which has member that will point to underlying index that stores the data
 faiss::IndexIVFPQ * extractIVFPQIndex(faiss::Index * index);
 
+void knn_jni::faiss_wrapper::CreateBinaryIndex(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jintArray idsJ, jlong vectorsAddressJ, jint dimJ,
```
There is a corresponding method in Java side which is FaissService -> JNIService. We are adding unit test there.

heemin32 commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -409,6 +505,114 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex_WithFilter(knn_jni::JNIUtilInter
     return results;
 }
 
+jobjectArray knn_jni::faiss_wrapper::QueryBinaryIndex_WithFilter(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jlong indexPointerJ,
```
There is a java method which linked to this method and we can write a unit test in java side. Is there any reason why you want to write unit test in cpp side?

heemin32 commented on jni/src/jni_util.cpp:
```diff
@@ -302,6 +335,23 @@ int knn_jni::JNIUtil::GetInnerDimensionOf2dJavaFloatArray(JNIEnv *env, jobjectAr
     return dim;
 }
 
+int knn_jni::JNIUtil::GetInnerDimensionOf2dJavaByteArray(JNIEnv *env, jobjectArray array2dJ) {
```
Where do you see `index` in the signature?

heemin32 replied:
Yes. This method has an assumption that the length of element in the array is all same.
Still, opening up the index is unnecessary at this point imo.

heemin32 commented on src/main/java/org/opensearch/knn/jni/JNIService.java:
```diff
@@ -50,7 +52,11 @@ public static void createIndex(
 
         if (KNNEngine.FAISS == knnEngine) {
-            FaissService.createIndex(ids, vectorsAddress, dim, indexPath, parameters);
+            if (parameters.get(KNNConstants.INDEX_DESCRIPTION_PARAMETER).toString().startsWith(FAISS_BINARY_INDEX_PREFIX)) {
```
index description is required in index factory. See
https://github.com/facebookresearch/faiss/wiki/The-index-factory
https://github.com/facebookresearch/faiss/wiki/Binary-indexes

Exact match does not work here.

shatejas replied:
Can't we have the next best thing?
```
Static set of these strings
BFlat 
BIVF1024,
BHNSW16 
BIVF1024_HNSW16
```

```
if(Set.contains(indexDescription)) ....
```

heemin32 replied:
No. The number is not fixed. User can provide any.

heemin32 commented on src/main/java/org/opensearch/knn/jni/JNIService.java:
```diff
@@ -50,7 +52,11 @@ public static void createIndex(
-            FaissService.createIndex(ids, vectorsAddress, dim, indexPath, parameters);
+            if (parameters.get(KNNConstants.INDEX_DESCRIPTION_PARAMETER).toString().startsWith(FAISS_BINARY_INDEX_PREFIX)) {
+                FaissService.createBinaryIndex(ids, vectorsAddress, dim, indexPath, parameters);
+            } else {
```
I don't think having faiss specific logic in java side is a bad idea. I would rather make cpp side as simple as possible and add most of logic in java side. It is a pain to work on cpp side tbh.

heemin32 commented on src/main/java/org/opensearch/knn/jni/JNIService.java:
```diff
@@ -195,6 +205,39 @@ public static KNNQueryResult[] queryIndex(
+     * @param filterIdsType how to filter ids: Batch or BitMap
+     * @return KNNQueryResult array of k neighbors
+     */
+    public static KNNQueryResult[] queryBinaryIndex(
```
There is no binary type. We pack the binary in byte type. Therefore the byte is just a place holder for 8 binary. It does not query byte.

heemin32 commented on src/test/java/org/opensearch/knn/jni/JNIServiceTests.java:
```diff
@@ -647,6 +648,24 @@ public void testCreateIndex_faiss_valid() throws IOException {
 
+    @SneakyThrows
+    public void testCreateIndex_binary_faiss_valid() {
+        List<String> methods = ImmutableList.of(faissBinaryMethod);
```
I can remove it. Just following the pattern of other tests.

heemin32 commented on src/test/java/org/opensearch/knn/jni/JNIServiceTests.java:
```diff
@@ -647,6 +648,24 @@ public void testCreateIndex_faiss_valid() throws IOException {
+                ImmutableMap.of(INDEX_DESCRIPTION_PARAMETER, method),
+                KNNEngine.FAISS
+            );
+            assertTrue(tmpFile1.toFile().length() > 0);
```
Agree that this is not a strong test case.
We should refactor the code to make those method as non static so the JNIService method is testable...

shatejas commented on jni/include/faiss_index_create.h:
```diff
@@ -0,0 +1,100 @@
+        void create() override;
+};
+
+class IndexCreateServiceFactory {
```
nit: formatting

heemin32 replied:
If you are referring the missing indentation with two namespaces, it is by intention to avoid having large space for each line.

shatejas commented on jni/include/faiss_util.h:
```diff
@@ -14,11 +14,29 @@
-#include "faiss/impl/IDGrouper.h"
+#include "jni_util.h"
+#include "faiss/impl/io.h"
+#include "faiss/index_factory.h"
+#include "faiss/index_io.h"
+#include "faiss/IndexHNSW.h"
+#include "faiss/IndexIVFFlat.h"
+#include "faiss/MetaIndexes.h"
+#include "faiss/Index.h"
+#include "faiss/impl/IDSelector.h"
+#include "faiss/IndexIVFPQ.h"
+#include "faiss/IndexBinaryIVF.h"
```
nit: wondering if these are needed

heemin32 replied:
Will check. Thanks.

shatejas commented on jni/src/faiss_index_create.cpp:
```diff
@@ -0,0 +1,265 @@
+    std::unordered_map<std::string, jobject> parameters
+) : jniUtil(jniUtil), env(env), metric(metric), indexDescription(indexDescription), dim(dim), numIds(numIds), threadCount(threadCount), vectorsAddress(vectorsAddress), ids(ids), indexPath(indexPath), parameters(parameters) {}
+
+void IndexCreateService::create() {
```
Thanks for separating this out. The code structure looks much better

navneet1v commented on jni/include/faiss_index_create.h:
```diff
@@ -0,0 +1,100 @@
+namespace knn_jni {
+namespace faiss_wrapper {
+
+class IndexCreateService {
```
My only question on this would be should we make this as IndexService rather than index create service so it can support both indexing and search. We can start with create functions for now. @heemin32 whats your thoughts?

heemin32 replied:
Yes. By looking at the code again, it is badly written. Instead of having all required values as class attributes, I can simply receive them. as create method parameter.

I think factory class is not needed and the code in factory class can be in JNI method. Let me refactor it.

navneet1v replied:
Sure.. Lets give this a try. Thanks for taking up this effort. This is really tricky and was pending for a long time. Appreciate all the effort we can put to reduce the complexity

heemin32 commented on src/main/java/org/opensearch/knn/jni/JNIService.java:
```diff
@@ -101,7 +103,12 @@ public static long loadIndex(String indexPath, Map<String, Object> parameters, K
 
         if (KNNEngine.FAISS == knnEngine) {
-            return FaissService.loadIndex(indexPath);
+            if (parameters.get(KNNConstants.INDEX_DESCRIPTION_PARAMETER) != null
+                && parameters.get(KNNConstants.INDEX_DESCRIPTION_PARAMETER).toString().startsWith(FAISS_BINARY_INDEX_PREFIX)) {
+                return FaissService.loadBinaryIndex(indexPath);
+            } else {
+                return FaissService.loadIndex(indexPath);
+            }
```
In my opinion, cpp code and JNI layer should be as thin as possible for better maintainability and development efficiency. 
Any code duplication in cpp code can be avoided if these two methods share single method internally.

navneet1v replied:
> Any code duplication in cpp code can be avoided if these two methods share single method internally.

Yes this is what my point was. So, main question is how difficult it is to say we create a functions called as


`FaissService.loadIndex(indexPath, indexType)`, here indexType can be a string or integer whatever suits us whose value can be binary or empty if not binary. Then in the cpp layer we check if the indexTypeString == 'binary' then we load the index as a binary index.

As you are making the change and more closer to the code then I am, would like to know if this is something we want to take right now if its a small change or it is big enough that we will need a separate PR.

heemin32 replied:
You are right. We make a decision in cpp for createIndex and make a decision in java for load/query index. Let me  move the decision role from cpp to java for create index method as well.

heemin32 replied:
> > Any code duplication in cpp code can be avoided if these two methods share single method internally.
> 
> Yes this is what my point was. So, main question is how difficult it is to say we create a functions called as
> 
> `FaissService.loadIndex(indexPath, indexType)`, here indexType can be a string or integer whatever suits us whose value can be binary or empty if not binary. Then in the cpp layer we check if the indexTypeString == 'binary' then we load the index as a binary index.
> 
> As you are making the change and more closer to the code then I am, would like to know if this is something we want to take right now if its a small change or it is big enough that we will need a separate PR.

Just passing indexType is easy work to do. It is more about our strategy on implementation; if we want add more logic to cpp or java. I think it is better to keep the logic as simple as possible in cpp side unless there are performance disadvantage.

navneet1v replied:
I agree in terms of keeping things as simple as possible. But I would like to have them at 1 place either in java layer or in cpp layer. In my true opinion I would like to have things in java layer as we get more control.

heemin32 replied:
Yes. Moved them to java layer.

heemin32 commented:
Refactored and added unit tests for createIndex method. For queryIndex, it needs another big change as current code shape is not test friendly. Hope, we can improve it incrementally outside of this PR.

navneet1v commented on jni/src/faiss_index_service.cpp:
```diff
@@ -0,0 +1,174 @@
+namespace knn_jni {
+namespace faiss_wrapper {
+
+template<typename INDEX, typename IVF, typename HNSW>
+void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
```
really good abstraction.. Just by looking at it feels overkill, but when I looked at the reusability of the code looks absolutely awesome. But not sure if this is the best way to achieve this.

heemin32 replied:
Yeah. It was my best effort. There could be better a way though.

navneet1v replied:
I think I am good on this for now. Till someone contribute a better way that this.

navneet1v commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -108,63 +110,61 @@ void knn_jni::faiss_wrapper::CreateIndex(knn_jni::JNIUtilInterface * jniUtil, JN
     // This is not the ideal approach, please refer this gh issue for long term solution:
     // https://github.com/opensearch-project/k-NN/issues/1600
-    delete inputVectors;
+    if (isBinaryIndex) {
+        delete reinterpret_cast<std::vector<uint8_t>*>(vectorsAddress);
+    }
+    else {
+        delete reinterpret_cast<std::vector<float>*>(vectorsAddress);
+    }
```
why we need to `reinterpret_cast` the casting here? We can just call delete and then the address?

heemin32 commented on jni/include/faiss_index_service.h:
```diff
@@ -0,0 +1,76 @@
+
+class IndexService {
+public:
+    IndexService(std::unique_ptr<FaissMethods> faissMethods);
```
For better testability. This help us to mock faiss method and write unit test easily.

navneet1v replied:
I would like to see if there is a better way to mock, not a blocker for now.

heemin32 commented on jni/include/faiss_index_service.h:
```diff
@@ -0,0 +1,76 @@
+namespace knn_jni {
+namespace faiss_wrapper {
+
+class IndexService {
```
ack

heemin32 commented on src/main/java/org/opensearch/knn/index/KNNIndexShard.java:
```diff
@@ -171,6 +179,16 @@ List<EngineFileContext> getEngineFileContexts(IndexReader indexReader, KNNEngine
                     SpaceType spaceType = SpaceType.getSpace(spaceTypeName);
                     String modelId = fieldInfo.attributes().getOrDefault(MODEL_ID, null);
+                    String indexDescription = null;
+                    if (KNNEngine.FAISS.equals(knnEngine)
+                        && fieldInfo.attributes().get(KNNConstants.KNN_ENGINE).equals(KNNEngine.FAISS.getName())) {
+                        indexDescription = (String) XContentHelper.createParser(
+                            NamedXContentRegistry.EMPTY,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                            new BytesArray(fieldInfo.attributes().get(KNNConstants.PARAMETERS)),
+                            MediaTypeRegistry.getDefaultMediaType()
+                        ).map().getOrDefault(INDEX_DESCRIPTION_PARAMETER, null);
```
ack

heemin32 commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -227,6 +233,16 @@ private Map<Integer, Float> doANNSearch(final LeafReaderContext context, final B
         }
 
+        String indexDescription = null;
+        if (KNNEngine.FAISS.equals(knnEngine) && modelId == null) {
```
ack

heemin32 commented on jni/include/faiss_index_service.h:
```diff
@@ -0,0 +1,76 @@
+class IndexService {
+public:
+    IndexService(std::unique_ptr<FaissMethods> faissMethods);
+    virtual void createIndex(
```
I didn't see it as an issue enough to be mentioned in TODO but I can add.

jmazanec15 replied:
Should we put parameter definitions?

shatejas commented on src/main/java/org/opensearch/knn/index/IndexUtil.java:
```diff
@@ -286,4 +303,13 @@ public static boolean isSharedIndexStateRequired(KNNEngine knnEngine, String mod
         return JNIService.isSharedIndexStateRequired(indexAddr, knnEngine);
     }
 
+    public static String getIndexDescription(FieldInfo fieldInfo) throws IOException {
```
nit: You are extracting information from FieldInformation here. this in indexUtil seems out of place. Maybe consider creating a class `FieldInfoExtractor` and having a method there? That class will have a defined purpose. Having a generic Util class has overloaded responsibilities, we should start moving away from it

shatejas commented on src/main/java/org/opensearch/knn/index/KNNIndexShard.java:
```diff
@@ -171,6 +173,11 @@ List<EngineFileContext> getEngineFileContexts(IndexReader indexReader, KNNEngine
                     SpaceType spaceType = SpaceType.getSpace(spaceTypeName);
                     String modelId = fieldInfo.attributes().getOrDefault(MODEL_ID, null);
+                    String indexDescription = null;
+                    if (KNNEngine.FAISS.equals(knnEngine)
```
nit: Is this `if` branching required? let it return null if it doesn't find it in the fieldinfo. You have to handle null case irrespective of the branching

heemin32 commented on jni/include/faiss_methods.h:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A class having wrapped faiss methods
+ */
+class FaissMethods {
```
Might not possible. They are just simple method but not method under struct or class.

jmazanec15 commented on jni/include/commons.h:
```diff
@@ -26,12 +26,34 @@ namespace knn_jni {
          */
         jlong storeVectorData(knn_jni::JNIUtilInterface *, JNIEnv *, jlong , jobjectArray, jlong);
 
+        /**
```
Should we specify the data type the bytes will be stored in?

heemin32 replied:
It is in the parameter definition.
`@param data 2D byte array containing data to be stored in native memory.`

jmazanec15 commented on jni/include/faiss_index_service.h:
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
```
Like this class!

jmazanec15 commented on jni/include/faiss_methods.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * This class helps to mock faiss methods during unit test
+ */
+class FaissMethods {
```
Are we missing some methods? i.e. load, free, search,...

heemin32 replied:
I haven't moved all methods yet. We will add incrementally.

jmazanec15 replied:
Lets add todo

heemin32 replied:
I didn't want to go through and see what is needed to be in this class to add todo. We can add methods as needed base. For example, we don't need to add search method here because search is a method of Index struct which can be mocked without this class.

heemin32 commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -409,6 +410,114 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex_WithFilter(knn_jni::JNIUtilInter
+        std::vector<uint64_t> idGrouperBitmap;
+        auto hnswReader = dynamic_cast<const faiss::IndexBinaryHNSW*>(indexReader->index);
+        if(hnswReader) {
+            // Setting the ef_search value equal to what was provided during index creation. SearchParametersHNSW has a default
+            // value of ef_search = 16 which will then be used.
+            hnswParams.efSearch = hnswReader->hnsw.efSearch;
+            hnswParams.sel = idSelector.get();
```
TODO might not be enough to catch it. Let me create GH issue instead. https://github.com/opensearch-project/k-NN/issues/1769

heemin32 commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -247,6 +235,19 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
     return (jlong) indexReader;
 }
 
+jlong knn_jni::faiss_wrapper::LoadBinaryIndex(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jstring indexPathJ) {
```
Yes. Didn't add it now because wasn't sure if current approach of using index service will be accepted.
Once this PR is merged, will move the load index method into index service in a separate PR.

jmazanec15 reviewed (commented):
I like the refactors and I think it looks good. One thing - can we run cpp unit tests with valgrind on linux to see if we are leaking any memory?

heemin32 commented:
Valgrind result
```
~/k-NN/jni$ valgrind --leak-check=yes ./bin/jni_test 
==54535== Memcheck, a memory error detector
==54535== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==54535== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==54535== Command: ./bin/jni_test
==54535== 
vex amd64->IR: unhandled instruction bytes: 0x62 0xF2 0xFD 0x8 0x7C 0xC0 0xC5 0xF9 0x7F 0x5
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==54535== valgrind: Unrecognised instruction at address 0x6a15c9e.
==54535==    at 0x6A15C9E: _GLOBAL__sub_I_distcomp_sparse_scalar_fast.cc (in /home/ubuntu/k-NN/jni/release/libopensearchknn_nmslib.so)
==54535==    by 0x400571E: call_init.part.0 (dl-init.c:74)
==54535==    by 0x4005823: call_init (dl-init.c:120)
==54535==    by 0x4005823: _dl_init (dl-init.c:121)
==54535==    by 0x401F59F: ??? (in /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2)
==54535== Your program just tried to execute an instruction that Valgrind
==54535== did not recognise.  There are two possible reasons for this.
==54535== 1. Your program has a bug and erroneously jumped to a non-code
==54535==    location.  If you are running Memcheck and you just saw a
==54535==    warning about a bad jump, it's probably your program's fault.
==54535== 2. The instruction is legitimate but Valgrind doesn't handle it,
==54535==    i.e. it's Valgrind's fault.  If you think this is the case or
==54535==    you are not sure, please let us know and we'll try to fix it.
==54535== Either way, Valgrind will now raise a SIGILL signal which will
==54535== probably kill your program.
==54535== 
==54535== Process terminating with default action of signal 4 (SIGILL)
==54535==  Illegal opcode at address 0x6A15C9E
==54535==    at 0x6A15C9E: _GLOBAL__sub_I_distcomp_sparse_scalar_fast.cc (in /home/ubuntu/k-NN/jni/release/libopensearchknn_nmslib.so)
==54535==    by 0x400571E: call_init.part.0 (dl-init.c:74)
==54535==    by 0x4005823: call_init (dl-init.c:120)
==54535==    by 0x4005823: _dl_init (dl-init.c:121)
==54535==    by 0x401F59F: ??? (in /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2)
==54535== 
==54535== HEAP SUMMARY:
==54535==     in use at exit: 83,256 bytes in 40 blocks
==54535==   total heap usage: 40 allocs, 0 frees, 83,256 bytes allocated
==54535== 
==54535== LEAK SUMMARY:
==54535==    definitely lost: 0 bytes in 0 blocks
==54535==    indirectly lost: 0 bytes in 0 blocks
==54535==      possibly lost: 0 bytes in 0 blocks
==54535==    still reachable: 83,256 bytes in 40 blocks
==54535==         suppressed: 0 bytes in 0 blocks
==54535== Reachable blocks (those to which a pointer was found) are not shown.
==54535== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==54535== 
==54535== For lists of detected and suppressed errors, rerun with: -s
==54535== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Illegal instruction (core dumped)
```

navneet1v commented on jni/src/faiss_index_service.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+}
+
+IndexService::IndexService(std::unique_ptr<FaissMethods> faissMethods) : faissMethods(std::move(faissMethods)) {}
```
The std::move is an interesting choice. But what my thinking is why we are using std::move here rather than just say a [unique pointer](https://en.cppreference.com/w/cpp/memory/unique_ptr) or [may do a call by reference](https://www.tutorialspoint.com/cplusplus/cpp_function_call_by_reference.htm) if we are concerned about copying of objects.

heemin32 replied:
Call by reference increase a chance of memory leak by bug.

navneet1v replied:
as this is an object, and you used the unique pointer, I doubt its the case. But if we 100% sure that this will not cause any performance issues and other memory leak I am okay to approve the code.

heemin32 replied:
This std::move is passing its ownership. If you don't use it, the ownership won't be transferred to IndexService and the memory can be released while IndexService sill needs it.

navneet1v commented on jni/src/faiss_wrapper.cpp:
```diff
@@ -409,6 +410,114 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex_WithFilter(knn_jni::JNIUtilInter
     return results;
 }
 
+jobjectArray knn_jni::faiss_wrapper::QueryBinaryIndex_WithFilter(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jlong indexPointerJ,
```
will the abstraction on the query flow be added in the upcoming PRs?

navneet1v replied:
Lets do that.

heemin32 replied:
https://github.com/opensearch-project/k-NN/issues/1773

navneet1v commented on src/main/java/org/opensearch/knn/jni/JNIService.java:
```diff
@@ -50,7 +52,12 @@ public static void createIndex(
 
         if (KNNEngine.FAISS == knnEngine) {
-            FaissService.createIndex(ids, vectorsAddress, dim, indexPath, parameters);
+            if (parameters.get(KNNConstants.INDEX_DESCRIPTION_PARAMETER) != null
```
This logic of identifying the binary index lets abstract this out in a utility. I think this is going to be used at other places too.

heemin32 replied:
As this is java side code, will make the change in the next PR.
